src/Voron/StorageEnvironmentOptions.cs

```diff
@@ -606,6 +605,50 @@ public override bool TryDeleteJournal(long number)
                 return true;
             }
 
+            public override unsafe bool ReadValidMetadata(string filename, out MetadataFile metadata)
+            {
+                metadata = default;
+                var path = _basePath.Combine(filename);
+                if (File.Exists(path.FullPath) == false)
+                {
+                    return false;
+                }
+
+                using (var fs = SafeFileStream.Create(path.FullPath, FileMode.Open, FileAccess.ReadWrite, FileShare.Read, 4096, FileOptions.None))
+                {
+                    var metadataBuf = stackalloc MetadataFile[1];
+
+                    Span<byte> sizeBuffer = stackalloc byte[sizeof(int)];
+                    fs.ReadExactly(sizeBuffer);
+                    var size = metadataBuf[0].DataSize = BitConverter.ToInt32(sizeBuffer);
+
+                    Span<byte> bodyBuffer = new ((byte*)metadataBuf + sizeof(int), size - sizeof(int));
+                    var totalRead = 0;
+                    while (totalRead < bodyBuffer.Length)
+                    {
+                        var read = fs.Read(bodyBuffer[totalRead..]);
+                        if (read == 0)
+                            break;
+                        totalRead += read;
+                    }
+
+                    ulong hash = Hashing.XXHash64.CalculateInline(bodyBuffer[sizeof(ulong)..]);
+                    if (metadataBuf->Hash != hash)
+                        return false;
+
+                    metadata = metadataBuf[0];
+                    return true;
+                }
+            }
+
+            public override unsafe void WriteMetadata(string filename, MetadataFile metadata)
+            {
+                var path = _basePath.Combine(filename);
+                var rc = Pal.rvn_write_header(path.FullPath, (byte*)&metadata, metadata.DataSize, out var errorCode);
```

Comment:
Verify that this does fsync (pretty sure it does, but still)

Reply:
it does `FlushFileBuffers`